src/DefinitionResolver.php

```diff
@@ -575,7 +575,7 @@ public function resolveExpressionNodeToType($expr)
         //   $myVariable -> type of corresponding assignment expression
         if ($expr instanceof Node\Expression\Variable || $expr instanceof Node\UseVariableName) {
             if ($expr->getName() === 'this') {
-                return new Types\This;
+                return new Types\Object_(new Fqsen('\\' . $this->getContainingClassFqn($expr)));
```

Comment:
Docblocks are parsed by phpDocumentor, which (as you can see in the deleted line) has support for a `$this` return type